mmda/types/indexers.py

```diff
@@ -21,31 +23,73 @@ class Indexer:
+    Builds a static nested containment list from SpanGroups
+    and accepts other SpanGroups as search probes.
+
+    See: https://github.com/biocore-ntnu/ncls
+
+    [citation]
+    Alexander V. Alekseyenko, Christopher J. Lee;
+    Nested Containment List (NCList): a new algorithm for accelerating interval query of genome
+      alignment and interval databases, Bioinformatics,
+    Volume 23, Issue 11, 1 June 2007, Pages 1386–1393, https://doi.org/10.1093/bioinformatics/btl647
+    """
+
+    def __init__(self, span_groups: List[SpanGroup]) -> None:
+        starts = []
+        ends = []
+        ids = []
+
+        for sg_id, span_group in enumerate(span_groups):
+            for span in span_group.spans:
+                starts.append(span.start)
+                ends.append(span.end)
+                ids.append(sg_id)
+
+        self._sgs = span_groups
+        self._index = NCLS(
+            pd.Series(starts, dtype=np.int64),
+            pd.Series(ends, dtype=np.int64),
+            pd.Series(ids, dtype=np.int64)
+        )
+
+        self._ensure_disjoint()
+
+    def _ensure_disjoint(self) -> None:
+        """
+        Constituent span groups must be fully disjoint.
+        Ensure the integrity of the built index.
+        """
+        for span_group in self._sgs:
+            for span in span_group.spans:
+                matches = [match for match in self._index.find_overlap(span.start, span.end)]
+                if len(matches) > 1:
+                    raise ValueError(
+                        f"Detected overlap with existing SpanGroup(s) {matches} for {span_group}"
+                    )
 
-    # careful; if write it as _index = IntervalTree(), all SpanGroupIndexers will share the same _index object
-    _index: IntervalTree = field(default_factory=IntervalTree)
-
-    # TODO[kylel] - maybe have more nullable args for different types of queryes (just start/end ints, just SpanGroup)
     def find(self, query: SpanGroup) -> List[SpanGroup]:
         if not isinstance(query, SpanGroup):
             raise ValueError(f'SpanGroupIndexer only works with `query` that is SpanGroup type')
 
-        all_matched_span_groups = dict()
+        if not query.spans:
+            return []
+
+        matched_ids = set()
 
         for span in query.spans:
-            for matched_span_group in self._index[span.start : span.end]:
-                object_id = id(matched_span_group.data)
-                all_matched_span_groups[object_id] = matched_span_group.data
+            for _start, _end, matched_id in self._index.find_overlap(span.start, span.end):
+                matched_ids.add(matched_id)
+
+        matched_span_groups = [self._sgs[matched_id] for matched_id in matched_ids]
 
-        # Sort these because matching logic above doesn't preserve order
-        return sorted(list(all_matched_span_groups.values()))
+        # Retrieval above doesn't preserve document order; sort here
```

Comment:
lgtm; can you add a comment for a TODO to look into returning the matched span_groups in the same order as they appear in `self._sgs`?  realizing those could be different

Reply:
sure